### PR TITLE
Implemented reCaptcha

### PR DESCRIPTION
**Pull request by gianmarcomennecozzi**

### CODE REVIEW

File: app/captcha.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	err = json.Unmarshal(body, gr)
+	if err != nil {
+		r.lastError = append(r.lastError, err.Error())
+		return false
+	}
+	if !gr.Success {
+		r.lastError = append(r.lastError, gr.ErrorCodes...)
+	}
+	return gr.Success
+}
```
Comment (mrturkmencom) [inline app/captcha.go:58]: Great !

File: app/config.go [hunk 2]
```diff
@@ -37,6 +37,11 @@ type APIConfig struct {
 		Username string `yaml:"username"`
 		Password string `yaml:"password"`
 	} `yaml:"admin"`
+	Captcha struct {
+		Enabled   bool   `yaml:"enabled"`
+		SiteKey   string `yaml:"site-key"`
+		SecretKey string `yaml:"secret-key"`
+	} `yaml:"captcha"`
```
Comment (mrturkmencom) [inline app/config.go:44]: Nice to have it on config ! 👍🏻

File: app/templates.go [hunk 3]
```diff
@@ -139,3 +141,39 @@ const tooManyRequestsHTMLTemplate = `
...
+			  <br/>
+			</form>
+			<script>
+				function verifyCaptcha() {
+					document.getElementById("myForm").submit();
+				}
+			</script>
+		  </body>
+		</html>	`, "50%", "50%", "50%", formActionURL, siteKey)
+}
```
Comment (mrturkmencom) [inline app/templates.go:179]: We needed for haaukins 😉

File: app/config.go (hunk 2 above)
Comment (gianmarcomennecozzi) [reply app/config.go:44]: yeah so we can disable it in case. that's also how i made the tests work

File: app/templates.go (hunk 3 above)
Comment (gianmarcomennecozzi) [reply app/templates.go:179]: Yeah i think we need it for haaukins too :D

File: app/api.go [hunk 4]
```diff
@@ -45,6 +46,24 @@ func (lm *LearningMaterialAPI) request(next http.Handler) http.HandlerFunc {
...
+					w.WriteHeader(http.StatusBadRequest) //todo make 404 not found page
+					w.Header().Set("Content-Type", "text/html; charset=utf-8")
+					_, _ = w.Write([]byte(getCaptchaPage(formActionURL, lm.conf.API.Captcha.SiteKey)))
+					return
+				}
+				authC := http.Cookie{Name: chalsEncoded, Value: r.FormValue("g-recaptcha-response"), Path: "/", MaxAge: 200}
+				http.SetCookie(w, &authC)
+			}
+		}
+
```
Comment (gianmarcomennecozzi) [inline app/api.go:66]: It's actually a bit weird. but that was the only possible option to implement it

File: app/templates.go (hunk 3 above)
Comment (mrturkmencom) [reply app/templates.go:179]: I have just created the issue on haaukins repo.

File: app/api.go (hunk 4 above)
Comment (gianmarcomennecozzi) [reply app/api.go:66]: the code first checks if the captcha is enabled, 
then it check for a cookie based on the requested challenges,
if the cookie exists just goes to the next request (waiting page),
if not check if the captcha is correct, if yes goes to the next request (waiting page),
if not just shows the recaptcha page

Comment (gianmarcomennecozzi) [reply app/api.go:66]: the cookie expires after 200 seconds

**Pull request merged**